_posts/2017-07-29-css-grids.md

```diff
@@ -0,0 +1,170 @@
+    height: 98%;
+    float: right;
+    background-color: crimson;
+}
+```
+
+---
+
+Grid creates, well, a grid inside a div element (Think of lines between the space in the image below). 
+
+![Grid Layout]({{ site.url }}/assets/css-grids/grid-layout.png)
+
+---
+
+We can decide how many rows and columns we create as well as the gaps between them.
+After this any child elements (we cant place any child of child elements, that would be too inception-y) can be formed to fit inside this grid, which eliminates thinking about how to design a site using dozens of wrappers and placing elements inside boxes inside boxes.
+
+```css
+.wrap {
+    border: 5px solid red;
+    height: 400px;
+    display: grid;
+    grid-template-columns: 1fr 1fr;
+    grid-template-rows: 100%;
+    grid-gap: 10px;
+}
+
+.main-content {
+    margin-left: 1%;
+    grid-area: 0/1;
+    height: 98%;
+    background-color: aqua;
+}
+
+.sidebar {
+    margin-right: 1%;
+    grid-area: 1/2;
+    height: 98%;
+    background-color: crimson;
+}
+
+```
+
+This does the same thing, what we are doing is making a grid with 2 columns and one row, then setting the columns size to `1fr` which is 1 fraction of the available space
+The Grids example above looks quite ugly as you can imagine working with column and row numbers in `grid-area` will get complex with even a few of each, so next we can look at naming the layout sections for ease of use. Below is a mockup site made with `grid-template-areas`
+
+![Site using Grids]({{ site.url }}/assets/css-grids/site-with-grids.png)
+
+And the CSS:
+
+```css
+:root {
+  --grid-gap: 5px;
+  --font-size: 1em;
+}
+p {
+  font-size: var(--font-size);
+  display: inline-block;
+}
+
+#wrapper {
+  display: grid;
+  grid-template-columns: 2fr 4fr 1fr;
+  grid-template-rows: 4em 12em auto 4em;
+  grid-template-areas:
+  "nav nav nav"
+  "hd hd hd"
+  "side cnt cnt"
+  ". ft ft";
+  grid-gap: var(--grid-gap);
+  height: 100%;
+}
+
+.header {
+  background-color: blue;
+  grid-area: hd;
+  margin: 0 0 0 0;
+}
+
+.sidebar {
+  background-color: pink;
+  grid-area: side;
+}
+
+.navbar {
+  background-color: skyblue;
+  grid-area: nav;
+  --font-size: 1.5em;
+}
+
+.content {
+  background-color: yellow;
+  grid-area: cnt;
+}
+
+.footer {
+  background-color: grey;
+  grid-area: ft;
+}
+
+```
+
+Using `grid-tempate-areas` we write down for each column and row, you can do it all on one line but how its down above feels nicer for me. Below is a picture of how the above `grid-template-areas` code works
+
+![Creating a Grid]({{ site.url }}/assets/css-grids/creating-a-grid.png)
+
+After this we use `grid-area` and instead of the grid numbers we just use the grid area "name" (the "." here is an example of blank). Now here is the great part, say we want to re-organize for mobile, since we are using em and auto and fractions its already going to shrink properly, but if we want to re-organize it all we have to do is re-arrange the `grid-template-areas` and voila
```

Comment:
This is all kind of one run-on sentence.

Consider:

> After this we use `grid-area`, and instead of the grid numbers we just use the grid area "name" (the "." here is an example of blank). Now here is the great part. For example, if we want to re-organize the layout for mobile, we are already using `em`, `auto` and fractions so it will already shrink properly. But in any case, if we want to re-organize it all we have to do is re-arrange the `grid-template-areas`: